srtcore/sync.cpp

```diff
@@ -0,0 +1,168 @@
+#include "sync.h"
+#include "udt.h"
+#include "srt_compat.h"
+
+#if defined(_WIN32)
+#define TIMING_USE_QPC
+#include "win/wintime.h"
+#include <sys/timeb.h>
+#elif defined(OSX) || (TARGET_OS_IOS == 1) || (TARGET_OS_TV == 1)
+#define TIMING_USE_MACH_ABS_TIME
+#include <mach/mach_time.h>
+#elif defined(_POSIX_MONOTONIC_CLOCK) && _POSIX_TIMERS > 0
+#define TIMING_USE_MONOTONIC_CLOCK
+#endif
+
+namespace srt
+{
+namespace sync
+{
+
+void rdtsc(uint64_t& x)
+{
+#ifdef IA32
+    uint32_t lval, hval;
+    // asm volatile ("push %eax; push %ebx; push %ecx; push %edx");
+    // asm volatile ("xor %eax, %eax; cpuid");
+    asm volatile("rdtsc" : "=a"(lval), "=d"(hval));
+    // asm volatile ("pop %edx; pop %ecx; pop %ebx; pop %eax");
+    x = hval;
+    x = (x << 32) | lval;
+#elif defined(IA64)
+    asm("mov %0=ar.itc" : "=r"(x)::"memory");
+#elif defined(AMD64)
+    uint32_t lval, hval;
+    asm("rdtsc" : "=a"(lval), "=d"(hval));
+    x = hval;
+    x = (x << 32) | lval;
+#elif defined(TIMING_USE_QPC)
+    // This function should not fail, because we checked the QPC
+    // when calling to QueryPerformanceFrequency. If it failed,
+    // the m_bUseMicroSecond was set to true.
+    QueryPerformanceCounter((LARGE_INTEGER*)&x);
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+    x = mach_absolute_time();
+#else
+    // use system call to read time clock for other archs
+    timeval t;
+    gettimeofday(&t, 0);
+    x = t.tv_sec * uint64_t(1000000) + t.tv_usec;
+#endif
+}
+
+int64_t get_cpu_frequency()
+{
+    int64_t frequency = 1; // 1 tick per microsecond.
+
+#if defined(TIMING_USE_QPC)
+
+    LARGE_INTEGER ccf; // in counts per second
+    if (QueryPerformanceFrequency(&ccf))
+        frequency = ccf.QuadPart / 1000000; // counts per microsecond
+
+#elif defined(TIMING_USE_MACH_ABS_TIME)
+
+    mach_timebase_info_data_t info;
+    mach_timebase_info(&info);
+    frequency = info.denom * int64_t(1000) / info.numer;
+
+#elif defined(IA32) || defined(IA64) || defined(AMD64)
+    uint64_t t1, t2;
+
+    rdtsc(t1);
+    timespec ts;
+    ts.tv_sec  = 0;
+    ts.tv_nsec = 100000000;
+    nanosleep(&ts, NULL);
+    rdtsc(t2);
+
+    // CPU clocks per microsecond
+    frequency = int64_t(t2 - t1) / 100000;
+#endif
+
+    return frequency;
+}
+
+static const int64_t s_cpu_frequency = get_cpu_frequency();
+
+} // namespace sync
+} // namespace srt
+
+template <>
+uint64_t srt::sync::TimePoint<srt::sync::steady_clock>::us_since_epoch() const
+{
+    return m_timestamp / s_cpu_frequency;
+}
+
+template <>
+srt::sync::Duration<srt::sync::steady_clock> srt::sync::TimePoint<srt::sync::steady_clock>::time_since_epoch() const
+{
+    return srt::sync::Duration<srt::sync::steady_clock>(m_timestamp);
+}
+
+srt::sync::TimePoint<srt::sync::steady_clock> srt::sync::steady_clock::now()
+{
+    uint64_t x = 0;
+    rdtsc(x);
+    return TimePoint<steady_clock>(x);
+}
+
+long long srt::sync::count_microseconds(const steady_clock::duration& t) { return t.count() / s_cpu_frequency; }
```

Comment:
For the sake of clarity, consistency, and aesthetics, could you please expand this and a couple of other functions below to be multiline?